pkg/chains/signing.go

```diff
@@ -252,10 +309,34 @@ func (ts *TaskRunSigner) SignTaskRun(ctx context.Context, tr *v1beta1.TaskRun) e
 			}
 		}
 		if merr.ErrorOrNil() != nil {
+			if err := HandleRetries(tr, ts, extraAnnotations); err != nil {
+				merr = multierror.Append(merr, err)
+			}
 			return merr
 		}
 	}
 
 	// Now mark the TaskRun as signed
 	return MarkSigned(tr, ts.Pipelineclientset, extraAnnotations)
 }
+
+// HandleRetries adds retry annotation to keep track of retries and marks the job as failed
+// if retries have reached MaxRetries.
+func HandleRetries(tr *v1beta1.TaskRun, ts *TaskRunSigner, extraAnnotations map[string]string) error {
+	ok := RetryAgain(tr)
```

Comment:
I think this function just has to say:

```go
	if RetryAgain(tr) {
		return AddRetries(tr, ts.Pipelineclientset, extraAnnotations)

	}
	// otherwise, mark as failed
	return MarkFailed(tr, ts.Pipelineclientset, extraAnnotations)
```